feat(profiling): Add support for Rust profiles ingestion

viglia opened the pull request:
This PR aims at supporting the ingestion of Rust profiles.

This was tested e2e by sending and ingesting a sample rust profile.

[Link](https://github.com/getsentry/sentry/pull/35461) to related `sentry` PR

Commit:
Add support to ingest Rust profiles

Commit:
Add changelog entry

phacops commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
 
+#[derive(Debug, Serialize, Deserialize)]
+struct RustProfile {
+    duration_ns: u128,
```
We can't store a `u128` in the DB. Can we receive a `u64` instead?

phacops commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct RustDebugMeta {
```
I don't think we need all this to be `pub`.

viglia replied:
yes, as I mentioned in a previous comment I'm turning all the struct private since for now I see no need to have them public 👍🏻

jjbayer commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct RustDebugMeta {
```
nit: I think we have a convention of declaring types before they are used, so in this case `RustDebugMeta` before `RustProfile`. Maybe @jan-auer can confirm whether we still maintain that convention.

jjbayer commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
+    platform: String,
+    architecture: String,
+    // string representation of Uuid v4
+    trace_id: String,
```
Any reason `trace_id`,`transaction_id` and `profile_id` are not of type `EventId`, like in `AndroidProfile`?

https://github.com/getsentry/relay/blob/d8f887a595a3ee7c78c16d4513ab86242c859f50/relay-server/src/utils/profile.rs#L82

Maybe we could even define keys used in multiple profile structs in a common helper struct: https://serde.rs/attr-flatten.html#factor-out-frequently-grouped-keys

viglia replied:
That's a good point. Thanks for bringing this up.
Changed all of them to `EventID` as I don't have specific reasons to keep them as String

jan-auer replied:
In theory, `EventId` is more lenient than our span and trace IDs. See https://github.com/getsentry/relay/blob/a1303ce8eeb9ff479c91fa6642f197ba4ab2ff70/relay-general/src/protocol/contexts.rs#L313-L314. In the event protocol we do not allow slashes.

You can keep them `EventId` here since being more lenient isn't a problem, though a more correct name may be better in the long term.

Commit:
Change timing fields from u128 to u64

Rust as_nanos returns u128. Unfortunately we can't store u128 into the DB, nor can Golang handle u128.
As a fix here some fields are refactored to hold u64 values.
That's enough to store the timing info we need without causing an overflow.

Commit:
Refactor trace_id, transactiond_id and profile_id types from String to EventID

Commit:
Moves RustDebugMeta type declaration before it's used

jjbayer commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct RustSampledProfile {
```
nit: I would apply the same declaration order to the rest of the structs in this file, i.e. `RustSampledProfile` before `RustProfile`, `RustFrame` before `RustSample`, etc.

jjbayer commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
+    duration_ns: u64,
+    platform: String,
+    architecture: String,
+    // string representation of Uuid v4
```
Are these comments still correct? `EventId` _is_ a UUID, right?

viglia replied:
yes it's a `UUID`. Correct. I'm sweeping off the comments 👍🏻

jjbayer commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
+    /// different address.
+    ///
+    /// Symbols and addresses in the native image are always relative to the start of the image and do not consider the preferred load address. It is merely a hint to the loader.
+    pub image_vmaddr: Option<String>,
```
Any reason this does not have `#[serde(default, skip_serializing_if = "Option::is_none")]`?

viglia replied:
marking this as resolved as I merged this type with `NativeDebugImage` shared with `ios`

jjbayer reviewed (approved):
Looks good to me! Just a few minor comments.

jan-auer commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
 
+#[derive(Debug, Deserialize, Serialize)]
+pub struct RustDebugMeta {
+    pub images: Vec<SymbolicDebugImage>,
```
Let's start adding some doc comments to the public structures and their fields. For now it's still pretty straight forward what these mean, though it will help us maintain this code base going forward.

phacops replied:
We should make all these structures private for now, we don't want to document these right now.

viglia replied:
yes, I've indeed changed all of them to private now. No need to have them public for now 👍🏻

jan-auer commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
+
+/// Represents a symbolic debug image.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct SymbolicDebugImage {
```
This is almost the same as `MachOImage`, and in fact symbolicator treats them identically. Could we make sure that there's only a single image type? Ideally, this becomes identical to [`relay_general::protocol::NativeDebugImage`](https://github.com/getsentry/relay/blob/7d0dc11b38d4e876d57ad699252544d8f145a2cd/relay-general/src/protocol/debugmeta.rs#L309). When comparing, note that only attributes annotated with `metastructure(required = "true")` are required by the event protocol, although you can be stricter here.

I'm not sure if the rest of the profiling infrastructure is prepared to keep the structures the same, though this would be the ideal case.

phacops replied:
We can keep the structures the same, we're all good down the line.

viglia replied:
marking this as resolved as I merged this type with `NativeDebugImage` shared with `ios`

jan-auer commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct SymbolicDebugImage {
+    #[serde(rename = "type")]
+    pub image_type: String,
```
There's an enum for [`ImageType`](https://github.com/getsentry/relay/blob/904625ced5a11408f5d59046c03989bb929ad20f/relay-server/src/utils/profile.rs#L183-L187) already in this file that should be extended with the values derived from [here](https://github.com/getsentry/relay/blob/7d0dc11b38d4e876d57ad699252544d8f145a2cd/relay-general/src/protocol/debugmeta.rs#L430-L435):
- `"macho"`
- `"elf"`
- `"pe"`

jan-auer commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
+    /// Unique debug identifier of the image.
+    ///
+    /// This is also called `debug_id`.
+    pub id: String,
```
Import from the `debugid` crate.

```suggestion
    #[serde(alias = "id")]
    pub debug_id: DebugId,
```

viglia replied:
marking this as resolved as I merged this type with `NativeDebugImage` shared with `ios`

jan-auer commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
+
+    /// Optional identifier of the code file.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub code_id: Option<String>,
```
Import from the `debugid` crate.

```suggestion
    pub code_id: Option<CodeId>,
```

viglia replied:
marking this as resolved as I merged this type with `NativeDebugImage` shared with `ios`

jan-auer commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
+pub struct RustSample {
+    pub frames: Vec<RustFrame>,
+    pub thread_name: String,
+    pub thread_id: u64,
```
The thread_id can be alphanumeric on some platforms, at least we also support strings in the event protocol. Are you certain you can constrain this to `u64`? See [`ThreadId`](https://github.com/getsentry/relay/blob/cb4af7b2c50c1513ea084a1f899e835166a5a7ae/relay-general/src/protocol/thread.rs#L11-L19). You could import and use this enum directly here.

viglia replied:
@jan-auer yes, because client-side I'm using `pprof-rs` for which `thread_id` is only defined as `u64`.

see [here](https://docs.rs/pprof/0.8.0/pprof/struct.Frames.html)

jan-auer commented on relay-server/src/utils/profile.rs:
```diff
@@ -287,6 +287,110 @@ pub fn parse_typescript_profile(item: &mut Item) -> Result<(), ProfileError> {
+pub struct RustSample {
+    pub frames: Vec<RustFrame>,
+    pub thread_name: String,
+    pub thread_id: u64,
```
Note that we call these `id` and `name` in the Event protocol. I'm just realizing that we've already diverged from that with cocoa profiles, though these are sample structures and not threads. There's no strong need to align these, though I wanted to point it out.

jan-auer reviewed (approved):
Approving this ahead of time, please see the comments below for minor changes on field names and types used.

Commit:
Remove outdated field comments

Commit:
Refactor to use a common shared struct (NativeDebugImage) between ios and rust for debug images
